src/kbmod/search/raw_image_eigen.cpp

```diff
@@ -0,0 +1,767 @@
+    if (status == 105) {
+      status = 0;
+      fits_open_file(&fptr, filename.c_str(), READWRITE, &status);
+      if (status == 0) {
+        fits_delete_file(fptr, &status);
+        fits_create_file(&fptr, filename.c_str(), &status);
+      }
+    }
+
+    // Create the primary array image (32-bit float array)
+    long dimensions[2];
+    dimensions[0] = width;
+    dimensions[1] = height;
+    fits_create_img(fptr, FLOAT_IMG, 2 /*naxis*/, dimensions, &status);
+    fits_report_error(stderr, status);
+
+    /* Write the array of floats to the image */
+    fits_write_img(fptr, TFLOAT, 1, get_npixels(), image.data(), &status);
+    fits_report_error(stderr, status);
+
+    // Add the basic header data.
+    fits_update_key(fptr, TDOUBLE, "MJD", &obstime, "[d] Generated Image time", &status);
+    fits_report_error(stderr, status);
+
+    fits_close_file(fptr, &status);
+    fits_report_error(stderr, status);
+  }
+
+
+  void RawImageEigen::append_layer_to_file(const std::string& filename) {
+    int status = 0;
+    fitsfile* f;
+
+    // Check that we can open the file.
+    if (fits_open_file(&f, filename.c_str(), READWRITE, &status)) {
+      fits_report_error(stderr, status);
+      throw std::runtime_error("Unable to open FITS file for appending.");
+    }
+
+    // This appends a layer (extension) if the file exists)
+    /* Create the primary array image (32-bit float array) */
+    long dimensions[2];
+    dimensions[0] = width;
+    dimensions[1] = height;
+    fits_create_img(f, FLOAT_IMG, 2 /*naxis*/, dimensions, &status);
+    fits_report_error(stderr, status);
+
+    /* Write the array of floats to the image */
+    fits_write_img(f, TFLOAT, 1, get_npixels(), image.data(), &status);
+    fits_report_error(stderr, status);
+
+    // Save the image time in the header.
+    fits_update_key(f, TDOUBLE, "MJD", &obstime, "[d] Generated Image time", &status);
+    fits_report_error(stderr, status);
+
+    fits_close_file(f, &status);
+    fits_report_error(stderr, status);
+  }
+
+
+  // my implementation here, to test performance for this because no way the
+  // optimization from the original one actually can be worth the messiness
+  const float calc_median(std::vector<float>& values) {
+    std::sort(values.begin(), values.end());
+    const auto len = values.size();
+    return len % 2 == 0 ?
+      (values[len/2-1] + values[len/2]) / 2.0f :
+      values[len/2];
+  }
+
+
+  RawImageEigen create_median_image_eigen2(const std::vector<RawImageEigen>& images) {
+    int num_images = images.size();
+    assert(num_images > 0);
+
+    unsigned height = images[0].get_height();
+    unsigned width = images[0].get_width();
+    RawImageEigen result = RawImageEigen(height, width);
+
+    std::vector<float> pixels;
+    pixels.reserve(num_images);
+    for (unsigned y=0; y<height; ++y){
+      for (unsigned x=0; x<width; ++x){
+        int n_unmasked = 0;
+        for (int i=0; i<num_images; ++i){
```

Comment:
You are going to need to reset `pixels` outside of this loop or else it will grow as height * width * num_images instead of just to num_images.

Reply:
Removed. 

The name of the function was `create_median_image_eigen2`. It was a test function. The code here isn't what's really proposed to be adopted to be honest, it was posted more so that you can see the issue of public vs private Eigen matrix and the limitations of having to provide our setters. Hopefully now that I reverted much of it back to the original you'll find it more palatable.